PHPLIB-694: Spec tests for snapshot sessions

jmikola opened the pull request:
https://jira.mongodb.org/browse/PHPLIB-694

Syncs tests with mongodb/specifications/pull/1046 (will be updated once merged).

Temporarily reduces the `driver-versions` axis in the Evergreen matrix to `latest-dev` to pull in changes from [PHPC-1761](https://jira.mongodb.org/browse/PHPC-1761).

alcaeus commented on .evergreen/config.yml:
```diff
@@ -574,7 +574,7 @@ buildvariants:
 # Only tests against latest MongoDB and ext-mongodb versions
 - matrix_name: "test-php-versions"
-  matrix_spec: {"os-php7": "*", "php-versions": "*", "edge-versions": "latest-stable", "driver-versions": "latest-stable" }
+  matrix_spec: {"os-php7": "*", "php-versions": "*", "edge-versions": "latest-stable", "driver-versions": "latest-dev" }
```
I added todo comments in the lines where I changed the `matrix_spec` to ensure we're changing it back everywhere. Happy to skip this in favour of reverting this commit if the changes are isolated.

jmikola replied:
This will likely get squashed so I'll leave todo comments.

```suggestion
  # TODO: restore driver-versions to latest-stable after 1.11 is tagged
  matrix_spec: {"os-php7": "*", "php-versions": "*", "edge-versions": "latest-stable", "driver-versions": "latest-dev" }
```

alcaeus reviewed (approved):
LGTM, good to merge once tests have been updated.

tanlisu commented:
What is a prose test as opposed to a spec test (referred to in the [JIRA ticket](https://jira.mongodb.org/browse/PHPLIB-694))?

jmikola commented:
> What is a prose test as opposed to a spec test?

Spec tests are written in YML/JSON and executed using a test runner. [Prose](https://en.wikipedia.org/wiki/Prose) tests are expressed in language and implemented manually. The corresponding prose test for this work was defined in https://github.com/mongodb/specifications/blob/master/source/sessions/tests/README.rst#prose-tests and implemented in https://github.com/mongodb/mongo-php-driver/commit/ac4935338a65d7f3370befd5c245e1be68f9eeba#diff-f526132129853b0d00c150317c2cb972ac2d17d32e05fed83a6313d91fadcecf

jmikola commented on .evergreen/config.yml:
```diff
@@ -600,7 +600,7 @@ buildvariants:
 # Tests against various topologies
 - matrix_name: "test-mongodb-versions"
-  matrix_spec: {"os-php7": "rhel70-test", "php-edge-versions": "latest-stable", "versions": "*", "driver-versions": "latest-stable" }
+  matrix_spec: {"os-php7": "rhel70-test", "php-edge-versions": "latest-stable", "versions": "*", "driver-versions": "latest-dev" }
```
```suggestion
  # TODO: restore driver-versions to latest-stable after 1.11 is tagged
  matrix_spec: {"os-php7": "rhel70-test", "php-edge-versions": "latest-stable", "versions": "*", "driver-versions": "latest-dev" }
```

jmikola commented on .evergreen/config.yml:
```diff
@@ -611,22 +611,22 @@ buildvariants:
 # TODO: driver-versions can be changed back to lowest-supported when that version is enabled in the axis
 - matrix_name: "test-dependencies"
-  matrix_spec: { "dependencies": "lowest", "os-php7": "rhel70-test", "php-edge-versions": "oldest-supported", "versions": "*", "driver-versions": "latest-stable" }
+  matrix_spec: { "dependencies": "lowest", "os-php7": "rhel70-test", "php-edge-versions": "oldest-supported", "versions": "*", "driver-versions": "latest-dev" }
```
```suggestion
  # TODO: restore driver-versions to latest-stable after 1.11 is tagged
  matrix_spec: { "dependencies": "lowest", "os-php7": "rhel70-test", "php-edge-versions": "oldest-supported", "versions": "*", "driver-versions": "latest-dev" }
```

jmikola commented:
LGTMs noted. Will leave this open until https://github.com/mongodb/specifications/pull/1046 is merged and I can update the commit reference.
